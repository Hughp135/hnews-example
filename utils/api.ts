export const getTopStoryIDs = async (): Promise<number[]> => {
  const res = await fetch(
    "https://hacker-news.firebaseio.com/v0/topstories.json"
  );

  if (!res.ok) {
    throw new Error("Failed to fetch top stories");
  }

  return res.json();
};

export interface Story {
  by: string;
  score: number;
  time: number;
  title: string;
  type: string;
  url: string;
}

export const getStory = async (id: number): Promise<Story> => {
  const res = await fetch(
    `https://hacker-news.firebaseio.com/v0/item/${id}.json`
  );

  if (!res.ok) {
    throw new Error("Failed to fetch story data");
  }

  return res.json();
};
