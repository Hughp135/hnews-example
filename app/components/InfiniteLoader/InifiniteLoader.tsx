"use client";

import InfiniteScroll from "react-infinite-scroll-component";
import { useState } from "react";
import { StoryItem } from "../StoryItem/StoryItem";
import { getStory, Story } from "@/utils/api";
import { Spinner } from "../Spinner/Spinner";

const ITEMS_PER_PAGE = 16;

export const InfiniteLoader = ({ storyIds }: { storyIds: number[] }) => {
  const [numberOfItems, setNumberOfItems] = useState(0);
  const [stories, setStories] = useState<Story[]>([]);

  const loadMore = async () => {
    const idsToFetch = storyIds.slice(
      numberOfItems,
      numberOfItems + ITEMS_PER_PAGE
    );

    setNumberOfItems(numberOfItems + ITEMS_PER_PAGE);

    const newStories = await Promise.all(
      idsToFetch.map((storyId) => getStory(storyId))
    );

    setStories([...stories, ...newStories]);
  };

  return (
    <>
      <InfiniteScroll
        dataLength={stories.length}
        next={loadMore}
        hasMore={numberOfItems < storyIds.length}
        endMessage={<h4>No more stories to load</h4>}
        loader={<Spinner />}
        scrollThreshold={0.8}
      >
        {stories.map((story) => (
          <StoryItem key={story.time} story={story} />
        ))}
      </InfiniteScroll>
    </>
  );
};
