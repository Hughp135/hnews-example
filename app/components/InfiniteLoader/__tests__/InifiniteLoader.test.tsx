import { getStory, Story } from "@/utils/api";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { InfiniteLoader } from "../InifiniteLoader";

jest.mock("@/utils/api");

describe("InfiniteLoader", () => {
  const exampleStory: Story = {
    by: "test",
    score: 1,
    time: 201312,
    title: "A story is here",
    type: "story",
    url: "http://test.url",
  };

  beforeEach(() => {
    (getStory as jest.Mock).mockImplementation(async (storyId: number) => ({
      ...exampleStory,
      time: storyId,
    }));
  });

  it("should render no stories", () => {
    const { asFragment } = render(<InfiniteLoader storyIds={[]} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the loading state", () => {
    const ids = Array.from({ length: 100 }).map((_, idx) => idx);
    const { asFragment } = render(<InfiniteLoader storyIds={ids} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the first 16 stories after a scroll event occurs", async () => {
    const ids = Array.from({ length: 100 }).map((_, idx) => idx);
    const { asFragment } = render(<InfiniteLoader storyIds={ids} />);

    fireEvent.scroll(window, { target: { scrollY: 100 } });

    // Wait for new stories to be rendered
    await waitFor(
      async () =>
        expect(screen.queryAllByText("A story is here")).toHaveLength(16),
      {
        timeout: 100,
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 32 stories if scrolled twice", async () => {
    const ids = Array.from({ length: 100 }).map((_, idx) => idx);
    const { asFragment } = render(<InfiniteLoader storyIds={ids} />);

    fireEvent.scroll(window, { target: { scrollY: 100 } });

    // Wait for stories to be rendered
    await waitFor(
      () => expect(screen.queryAllByText("A story is here")).toHaveLength(16),
      {
        timeout: 500,
      }
    );

    fireEvent.scroll(window, { target: { scrollY: 100 } });

    // Wait for next stories to be rendered
    await waitFor(
      () => expect(screen.queryAllByText("A story is here")).toHaveLength(32),
      {
        timeout: 500,
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
