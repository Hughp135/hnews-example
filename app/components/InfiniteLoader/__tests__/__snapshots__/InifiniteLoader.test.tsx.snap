// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteLoader should render 32 stories if scrolled twice 1`] = `
<DocumentFragment>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component "
      style="height: auto; overflow: auto;"
    >
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InfiniteLoader should render no stories 1`] = `
<DocumentFragment>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component "
      style="height: auto; overflow: auto;"
    >
      <h4>
        No more stories to load
      </h4>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InfiniteLoader should render the first 16 stories after a scroll event occurs 1`] = `
<DocumentFragment>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component "
      style="height: auto; overflow: auto;"
    >
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
      <a
        class="container"
        href="http://test.url"
        rel="noreferrer"
        target="_blank"
      >
        <div
          class="container"
        >
          <article>
            <h3>
              A story is here
            </h3>
            <p>
              1 votes | test
            </p>
          </article>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InfiniteLoader should render the loading state 1`] = `
<DocumentFragment>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component "
      style="height: auto; overflow: auto;"
    >
      <div
        class="container"
      >
        <div
          class="spinner"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
